pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -1084,20 +1089,45 @@ func (a *Actuator) getNodeByMachine(ctx context.Context, machine *machinev1beta1
 remediateIfNeeded will try to remediate unhealthy machines (annotated by MHC) by power-cycle the host
 The full remediation flow is:
 1) Power off the host
-2) Add poweredOffForRemediation annotation to the unhealthy Machine
-3) Delete the node
-4) Power on the host
-5) Wait for the node the come up (by waiting for the node to be registered in the cluster)
-6) Restore node's annotations and labels
-7) Remove poweredOffForRemediation annotation, MAO's machine unhealthy annotation and annotations/labels backup
+2) Store node's annotations and labels
+3) Delete node
+4) Add poweredOffForRemediation annotation to the unhealthy Machine
+5) Power on the host
+6) Wait for the node the come up (by waiting for the node to be registered in the cluster)
+7) If the node doesn't come up after some time, and the host can be reprovisioned, delete the machine
+8) Restore node's annotations and labels
+9) Remove all remediation related annotations from the machine
 */
 func (a *Actuator) remediateIfNeeded(ctx context.Context, machine *machinev1beta1.Machine, baremetalhost *bmh.BareMetalHost) error {
 	if len(machine.Annotations) == 0 {
 		return nil
 	}
 
 	if _, needsRemediation := machine.Annotations[externalRemediationAnnotation]; !needsRemediation {
-		return nil
+		// It can happen that an unhealthy machine with external remediation annotation got healthy, and the external
+		// remediation annotation was removed by the MHC controller, before remediation finished or even started
+		// (e.g. the metal3 pod is not running, but machine was rebooted manually).
+		// In this case stop remediation, if possible.
+
+		// If bmh isn't requested to power off (1st step of remediation), and the machine isn't marked as powered off
+		// for remediation (removed as last step of remediation), there is no ongoing remediation
+		powerOffRequestExists := hasPowerOffRequestAnnotation(baremetalhost)
+		_, poweredOffForRemediationExists := machine.Annotations[poweredOffForRemediation]
+		if !powerOffRequestExists && !poweredOffForRemediationExists {
+			return nil
+		}
+
+		// We want to cancel a remediation, if nothing happened yet: power off was requested, but not executed yet.
+		if powerOffRequestExists && baremetalhost.Status.PoweredOn {
+			err := a.requestPowerOn(ctx, machine, baremetalhost)
+			// no need to requeue
+			if _, ok := err.(*machineapierrors.RequeueAfterError); ok {
+				return nil
+			}
+			return err
+		}
+
+		// In all other cases: there is an ongoing remediation, which should not be interrupted
```

Comment:
So if we find the power is already off here then we continue with the remediation, i.e. we ensure the Node is deleted, so that workloads fail over ASAP. In general this is good.

However, there is a race inherent in this. When we observe that the Host power is off, this is necessarily a view into the past. There is no way to guarantee that that Host will not get powered off, or even to know that it has not been powered off, and if it is we will not clean up the Node.

Reply:
if such race would occur, the host will be power-cycled, and from there MHC will re-evaluate if the node is healthy or not. 
if it is - we're good.
if it isn't, MHC will annotate the machine again and a new remediation flow will start, which will end up in cleaning the node.